apps/remote_control/lib/lexical/remote_control/search/store/backends/cub/state.ex

```diff
@@ -0,0 +1,242 @@
+  """
+  alias Lexical.Project
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+  alias Lexical.VM.Versions
+
+  import Record
+
+  defrecord :path_key, sort_key: 1, path: nil
+
+  defrecord :subject_key, type: nil, subtype: nil, subject: nil
+
+  defrecord :ref_key, id: nil
+
+  defstruct project: nil, leader?: false, leader_pid: nil, cub_pid: nil
+
+  # Erlang's total ordering
+  # number < atom < reference < fun < port < pid < tuple < map < nil < list < bit string
+
+  def new_leader(%Project{} = project, cub_pid) do
+    %__MODULE__{
+      project: project,
+      leader_pid: self(),
+      leader?: true,
+      cub_pid: cub_pid
+    }
+  end
+
+  def new_follower(%Project{} = project, leader_pid) do
+    %__MODULE__{
+      leader_pid: leader_pid,
+      leader?: false,
+      project: project
+    }
+  end
+
+  def prepare(%__MODULE__{leader?: true} = state) do
+    case CubDB.size(state.cub_pid) do
+      0 -> {:ok, :empty}
+      _n -> {:ok, :stale}
+    end
+  end
+
+  def prepare(%__MODULE__{}) do
+    {:ok, :stale}
+  end
+
+  def insert(%__MODULE__{} = state, entries) do
+    CubDB.put_multi(state.cub_pid, entries_to_keys_and_values(entries))
+  end
+
+  def drop(%__MODULE__{} = state) do
+    CubDB.clear(state.cub_pid)
+  end
+
+  def replace_all(%__MODULE__{} = state, entries) do
+    keys_and_values = entries_to_keys_and_values(entries)
+
+    CubDB.transaction(state.cub_pid, fn transaction ->
+      CubDB.Tx.clear(transaction)
+
+      transaction =
+        Enum.reduce(keys_and_values, transaction, fn {key, value}, transaction ->
+          CubDB.Tx.put(transaction, key, value)
+        end)
+
+      {:commit, transaction, :ok}
+    end)
+  end
+
+  def delete_by_path(%__MODULE__{} = state, path) do
+    CubDB.transaction(state.cub_pid, fn transaction ->
+      path_key = path_key(path: path)
+
+      {subject_keys, ref_keys} =
+        transaction
+        |> CubDB.Tx.get(path_key, [])
+        |> Enum.unzip()
+
+      deleted_entries =
+        ref_keys
+        |> Enum.map(&CubDB.Tx.get(transaction, &1))
+        |> Enum.to_list()
+
+      keys_to_delete = [path_key | subject_keys ++ ref_keys]
+
+      transaction =
+        Enum.reduce(keys_to_delete, transaction, fn key, tx ->
+          CubDB.Tx.delete(tx, key)
+        end)
+
+      {:commit, transaction, {:ok, deleted_entries}}
+    end)
+  end
+
+  def find_by_subject(%__MODULE__{} = state, type, subtype, subject) do
+    versions = Versions.current()
+
+    CubDB.with_snapshot(state.cub_pid, fn snap ->
+      snap
+      |> CubDB.Snapshot.select(min_key: min_subject_key(), max_key: max_subject_key())
+      |> Stream.flat_map(fn
+        {subject_key(subject: db_subject), entries} ->
+          Enum.filter(entries, fn entry ->
+            correct_version_range?(entry, versions) and
+              subject_matches?(db_subject, subject) and
+              type_and_subtype_match?(entry, type, subtype)
+          end)
+      end)
+      |> Enum.to_list()
```

Comment:
As is, this could be simplified a bit:

```suggestion
      |> Enum.flat_map(fn {subject_key(subject: db_subject), entries} ->
        Enum.filter(entries, fn entry ->
          correct_version_range?(entry, versions) and
            subject_matches?(db_subject, subject) and
            type_and_subtype_match?(entry, type, subtype)
        end)
      end)
```

However, do we need `type_and_subtype_match?(entry, type, subtype)`? If we have a `type` and `subtype` already, can't we use that in the `select` so that we only have to check versions and subject?

Reply:
Selects don't guarantee the subtype and types are always what you want. If the type is a wildcard, then you can get any subtype.

Reply:
The unit tests caught this

Reply:
Also, while reading the docs, it seemed to me that keeping filtering operations inside streams would be faster.